src/engine/sidechain/enginenetworkstream.cpp

```diff
@@ -0,0 +1,184 @@
+    }
+    int copyCount = math_min(writeAvailable, writeRequired);
+    if (copyCount > 0) {
+        (void)m_pOutputFifo->write(buffer, copyCount);
+        // we advance the frame only by the samples we have actually copied
+        // This means in case of buffer full (where we loose some frames)
+        // we do not get out of sync, and the syncing code tries to catch up the
+        // stream by writing silence, once the buffer is free.
+        m_streamFramesWritten += copyCount / m_numOutputChannels;
+    }
+    scheduleWorker();
+}
+
+void EngineNetworkStream::writeSilence(int frames) {
+    //qDebug() << "EngineNetworkStream::writeSilence()" << frames;
+    if (!m_pWorker->threadWaiting()) {
+        // no thread waiting, so we can advance the stream without
+        // buffering
+        m_streamFramesWritten += frames;
+        return;
+    }
+    int writeAvailable = m_pOutputFifo->writeAvailable();
+    int writeRequired = frames * m_numOutputChannels;
+    if (writeAvailable < writeRequired) {
+        qDebug() << "EngineNetworkStream::write() buffer full";
+    }
+    int clearCount = math_min(writeAvailable, writeRequired);
+    if (clearCount > 0) {
+        CSAMPLE* dataPtr1;
+        ring_buffer_size_t size1;
+        CSAMPLE* dataPtr2;
+        ring_buffer_size_t size2;
+        (void)m_pOutputFifo->aquireWriteRegions(clearCount,
+                &dataPtr1, &size1, &dataPtr2, &size2);
+        SampleUtil::clear(dataPtr1,size1);
+        if (size2 > 0) {
+            SampleUtil::clear(dataPtr2,size2);
+        }
+        m_pOutputFifo->releaseWriteRegions(clearCount);
+
+        // we advance the frame only by the samples we have actually cleared
+        m_streamFramesWritten += clearCount / m_numOutputChannels;
+    }
+    scheduleWorker();
+}
+
+void EngineNetworkStream::scheduleWorker() {
+    if (m_pOutputFifo->readAvailable()
+            >= m_numOutputChannels * kNetworkLatencyFrames) {
+        m_pWorker->outputAvailabe();
+    }
+}
+
+void EngineNetworkStream::read(CSAMPLE* buffer, int frames) {
+    int readAvailable = m_pOutputFifo->readAvailable();
+    int readRequired = frames * m_numInputChannels;
+    int copyCount = math_min(readAvailable, readRequired);
+    if (copyCount > 0) {
+        (void)m_pOutputFifo->read(buffer, copyCount);
+        buffer += copyCount;
+    }
+    if (readAvailable < readRequired) {
+        // Fill missing Samples with silence
+        int silenceCount = readRequired - readAvailable;
+        qDebug() << "EngineNetworkStream::write flushed" << readRequired
+                 << "samples";
+        SampleUtil::clear(buffer, silenceCount);
+    }
+}
+
+qint64 EngineNetworkStream::getStreamTimeFrames() {
+    return static_cast<double>(getStreamTimeUs()) * m_sampleRate / 1000000.0;
+}
+
+qint64 EngineNetworkStream::getStreamTimeUs() {
+    return getNetworkTimeUs() - m_streamStartTimeUs;
+}
+
+// static
+qint64 EngineNetworkStream::getNetworkTimeUs() {
```

Comment:
We should use a better way to keep time -- even if it doesn't match what libshout uses.

Changing the clock, daylight savings time, NTP updates, etc. can cause jumps forward or backward in time relative to ```m_streamStartTimeUs```. Changing your clock back in time before m_streamStartTimeUs will cause us to stop writing to libshout based on the current state of this PR.

We already have PerformanceTimer (src/util/performancetimer.h) and there is QElapsedTimer which both use monotonic clocks to ensure they aren't affected by these issues. I'm pretty surprised libshout uses gettimeofday!